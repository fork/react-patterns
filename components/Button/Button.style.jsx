import styled, { css } from 'styled-components';

import { btnReset, space, color, variant } from '../../stylesheets';

const StyledButton = styled.button`
  ${btnReset()}
  display: inline-block;
  font-size: 12px;

  /* Variant Prop */
  ${variant({
    primary: css`
      background: ${color('primary')};
      color: ${color('black')};
    `,
    secondary: css`
      background: ${color('black')};
      color: ${color('white')};
    `
  })}

  /* Size Prop */
  ${variant({
    prop: 'size',
    small: css`
      padding: ${space('s')} ${space('m')};
    `,
    large: css`
      padding: ${space('m')} ${space('l')};
    `
  })}
`;

export default StyledButton;
