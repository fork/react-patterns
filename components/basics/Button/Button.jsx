import React from 'react';
import PropTypes from 'prop-types';

import StyledButton from './Button.style';

const Button = ({ children, type, disabled, variant, size, href, ...props }) => (
  <StyledButton
    as={href ? 'a' : 'button'}
    href={href}
    type={href ? null : type}
    disabled={disabled}
    variant={variant}
    size={size}
    {...props}
  >
    {children}
  </StyledButton>
);

Button.propTypes = {
  children: PropTypes.node,
  disabled: PropTypes.bool,
  variant: PropTypes.oneOf(['primary', 'secondary']),
  size: PropTypes.oneOf(['small', 'large']),
  type: PropTypes.oneOf(['submit', 'button', 'reset']),
  /**
   * Defines the tag of the component. 'a' or 'button'.
   */
  href: PropTypes.string
};

Button.defaultProps = {
  children: null,
  disabled: false,
  type: 'button',
  size: 'small',
  variant: 'primary',
  href: null
};

export default Button;
