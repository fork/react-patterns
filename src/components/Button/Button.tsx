import React, { ReactNode } from 'react';

import StyledButton from './Button.style';

export type ButtonProps = {
  children: ReactNode;
  type?: 'button' | 'submit';
  disabled?: boolean;
  variant?: 'primary' | 'secondary';
  size?: 'small' | 'large';
  onClick?: (ev: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
};

const Button: React.FC<ButtonProps> = ({
  children,
  type = 'button',
  disabled,
  variant = 'primary',
  size = 'small',
  onClick
}: ButtonProps) => (
  <StyledButton type={type} disabled={disabled} variant={variant} size={size} onClick={onClick}>
    {children}
  </StyledButton>
);

export default Button;
