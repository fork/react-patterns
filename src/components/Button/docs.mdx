import { Story, Canvas } from '@storybook/addon-docs';

# Button

Buttons sind klickbare Elementen, mit denen Aktionen ausgelöst werden.

Buttons **werden verwendet**, wenn ein:e Nutzer:in:

- ein Formular absenden möchte
- eine neue Aufgabe/Aktion beginnen möchte

**Nicht eingesetzt** werden Buttons als navigatorische Elemente. Soll ein:e Nutzer:in auf eine neue Seite geleitet werden, wird stattdessen ein [Link](?path=/docs/components-navigation-link) verwendet.

## Varianten

Jede Button-Variante hat eine bestimmte Funktion, die durch das Design signalisiert wird. Daher ist es wichtig, dass die verschiedenen Varianten in Produkten konsistent eingesetzt werden, um Aktionen einheitlich zu kommunizieren.

Ein **Primary Button** wird für die wichtigste Aktion verwendet. Es soll nur ein Primary Button pro Screen verwendet werden (mit Ausnahme von Header, Dialogen oder Panels).

**Secondary Button**s werden für sekundäre Aktionen verwendet. Sie werden immer in Verbindung mit einem Primary Button verwendet und stellen die negative Aktion (z.B. "Abbrechen" oder "Zurück") dar. Secondary Buttons sollen nicht alleinstehend verwendet werden.

**Tertiary Button**s können für weniger prominente Aktionen verwendet werden. Sie können im Kombination mit einem Primary Button oder eigenständig eingesetzt werden.

<Canvas>
  <Story id="components-input-button--primary" />
</Canvas>

<Canvas>
  <Story id="components-input-button--secondary" />
</Canvas>

<Canvas>
  <Story id="components-input-button--tertiary" />
</Canvas>

<Canvas>
  <Story id="components-input-button--with-icon" />
</Canvas>

<Canvas>
  <Story id="components-input-button--disabled" />
</Canvas>

<!--
## Anatomie

Hier kann der Aufbau eines Buttons beschrieben werden.
-->

<!--
## Größen

Gibt es Buttons in unterschiedlichen Größen? Cool! Hier wird beschrieben, welche Größe wann eingesetzt wird.
-->
