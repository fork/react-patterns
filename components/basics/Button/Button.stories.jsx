import React from 'react';

import Button from './Button';

export default {
  title: 'Basics/Button',
  component: Button
};

export const Small = () => (
  <>
    <Button variant="primary">Primary</Button>
    <Button variant="secondary">Secondary</Button>
  </>
);

export const Large = () => (
  <>
    <Button variant="primary" size="large">
      Primary
    </Button>
    <Button variant="secondary" size="large">
      Secondary
    </Button>
  </>
);

export const asLink = () => (
  <>
    <Button variant="primary" size="large" href="https://wwww.fork.de" target="_blank">
      Primary
    </Button>
  </>
);
