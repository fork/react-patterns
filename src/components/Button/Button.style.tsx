import styled, { css } from 'styled-components';

import { btnReset, space, color, variant } from '../../stylesheets';
import { ButtonProps } from './Button';

export type StyledButtonProps = Pick<ButtonProps, 'variant' | 'size' | 'iconPosition'>;

const StyledButton = styled.button<StyledButtonProps>`
  ${btnReset()}
  display: inline-flex;
  align-items: center;
  font-size: 12px;
  border-radius: ${space('xxs')};

  /* variant Prop */
  ${variant({
    primary: css`
      background: ${color('primary')};
      color: ${color('white')};
    `,
    secondary: css`
      background: ${color('secondary')};
      color: ${color('black')};
    `
  })}

  /* size Prop */
    ${variant({
    prop: 'size',
    small: css`
      padding: ${space('s')} ${space('m')};
    `,
    large: css`
      padding: ${space('m')} ${space('l')};
    `
  })};

  /* iconPosition Prop */
  ${variant({
    prop: 'iconPosition',
    after: css`
      flex-direction: row-reverse;
    `,
    only: css`
      border-radius: 999px;
      padding: ${space('s')};
    `
  })};
`;

export default StyledButton;
