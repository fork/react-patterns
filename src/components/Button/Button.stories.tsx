import React from 'react';
import { Story, Meta } from '@storybook/react';

import Button, { ButtonProps } from './Button';
import docs from './docs.mdx';

export default {
  title: 'Components/Input/Button',
  component: Button,
  parameters: {
    docs: {
      page: docs
    }
  }
} as Meta;

const Template: Story<ButtonProps> = args => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'Small Button',
  size: 'small',
  variant: 'primary'
};

export const Large = Template.bind({});
Large.args = {
  label: 'Large Button',
  size: 'large',
  variant: 'primary'
};

export const Secondary = Template.bind({});
Secondary.args = {
  label: 'Secondary Button',
  variant: 'secondary'
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  label: 'Primary Button',
  variant: 'primary',
  icon: 'Expand'
};

export const Disabled = Template.bind({});
Disabled.args = {
  label: 'Primary Button',
  variant: 'primary',
  disabled: true
};
