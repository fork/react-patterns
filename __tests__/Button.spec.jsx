import React from 'react';
import { shallow } from 'enzyme';

import Button from '../components/basics/Button/Button';

describe('Button', () => {
  test('Button should render correctly', () => {
    const component = shallow(<Button>Primary Button</Button>);

    expect(component).toMatchSnapshot();
  });
});
