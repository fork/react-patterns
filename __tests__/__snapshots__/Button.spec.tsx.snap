// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Button should render correctly 1`] = `
<styled.button
  iconPosition="before"
  size="small"
  type="button"
  variant="primary"
>
  <styled.p
    variant="copy"
  />
</styled.button>
`;

exports[`Button Secondary Button should render correctly 1`] = `
<styled.button
  iconPosition="before"
  size="small"
  type="button"
  variant="secondary"
>
  <styled.p
    variant="copy"
  />
</styled.button>
`;
